nixos/modules/config/zswap.nix

```
@@ -0,0 +1,120 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+
+  cfg = config.zswap;
+
+in
+
+{
+  ###### interface
+
+  options = {
+
+    zswap = {
+
+      enable = lib.mkOption {
+        default = false;
+        type = lib.types.bool;
+        description = ''
+          Enable zswap, a lightweight compressed cache for swap pages.
+          See [
+          https://www.kernel.org/doc/Documentation/vm/zswap.txt
+          ] (https://www.kernel.org/doc/Documentation/vm/zswap.txt)
+        '';
+      };
+
+      maxPoolPercent = lib.mkOption {
+        default = 20;
+        type = lib.types.int;
+        description = ''
+          The maximum percentage of memory that the compressed pool can occupy.
+        '';
+      };
+
+      compressor = lib.mkOption {
+        default = "zstd";
+        example = "lzo";
+        type =
+          with lib.types;
+          either (enum [
+            "842"
+            "lzo"
+            "lz4"
+            "lz4hc"
+            "zstd"
+          ]) str;
+        description = ''
+          Compression algorithm. `lzo` has good compression,
+          but is slow. `lz4` has bad compression, but is fast.
+          `zstd` is both good compression and fast.
+        '';
+      };
+
+      zpool = lib.mkOption {
+        default = "zsmalloc";
+        type =
+          with lib.types;
+          either (enum [
+            "zsmalloc"
+            "zbud"
+          ]) str;
+        description = ''
+          Zswap makes use of zpool for managing the compressed memory pool.
+          See [
+            https://docs.kernel.org/admin-guide/mm/zswap.html#design
+          ] (https://docs.kernel.org/admin-guide/mm/zswap.html#design).
+        '';
+      };
+
+      acceptThresholdPercent = lib.mkOption {
+        default = 90;
+        type = lib.types.int;
+        description = ''
+          Sets the threshold at which zswap would start accepting pages again
+          after it became full.
+        '';
+      };
+    };
+
+  };
+
+  config = {
+    # systemd service to configure zswap parameters dynamically
+    systemd.services.zswap-activate = lib.mkIf cfg.enable {
+      description = "Configure zswap parameters via sysfs";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "local-fs.target" ];
+      serviceConfig = {
+        Type = "oneshot";
+        ExecStart = "${pkgs.writeShellScript "zswap-enable" ''
```

```suggestion
      serviceConfig.Type = "oneshot";
      script = ''
```
`script`, which is a (roughly speaking) wrapper for `serviceConfig.ExecStart = writeShellScript`, already exist

---

updated with suggestion.